rest/api_test.go

```diff
@@ -5619,3 +5619,46 @@ func TestUptimeStat(t *testing.T) {
 	log.Printf("uptime1: %d, uptime2: %d", uptime1, uptime2)
 	assert.True(t, uptime1 < uptime2)
 }
+
+func TestUserXattrSyncFn(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("Couchbase buckets only")
+	}
+
+	rt := NewRestTester(t, &RestTesterConfig{
+		SyncFn: `function(doc, oldDoc, meta){if(meta.xattrs.myxattr !== undefined){channel(meta.xattrs.myxattr);}}`,
+		DatabaseConfig: &DbConfig{
+			UserXattrKey: "myxattr",
+		},
+	})
+	defer rt.Close()
+
+	gocbBucket, ok := base.AsGoCBBucket(rt.Bucket())
+	if !ok {
+		t.Skip("Test requires Couchbase Bucket")
+	}
+
+	var doc struct {
+		SyncData db.SyncData `json:"_sync"`
+	}
+	res := rt.SendAdminRequest("PUT", "/db/doc", "{}")
+	assertStatus(t, res, http.StatusCreated)
+
+	res = rt.SendAdminRequest("GET", "/db/_raw/doc", "")
+	assertStatus(t, res, http.StatusOK)
+	err := json.Unmarshal(res.BodyBytes(), &doc)
+	require.NoError(t, err)
+	assert.Equal(t, []string{}, doc.SyncData.Channels.KeySet())
+
+	_, err = base.WriteXattr(gocbBucket, "doc", "myxattr", "channel")
+	assert.NoError(t, err)
+
+	res = rt.SendAdminRequest("PUT", "/db/doc?rev="+doc.SyncData.CurrentRev, "{}")
+	assertStatus(t, res, http.StatusCreated)
+
+	res = rt.SendAdminRequest("GET", "/db/_raw/doc", "")
+	assertStatus(t, res, http.StatusOK)
+	err = json.Unmarshal(res.BodyBytes(), &doc)
+	require.NoError(t, err)
+	assert.Equal(t, []string{"channel"}, doc.SyncData.Channels.KeySet())
```

Comment:
Test coverage for the following would be useful:
- updating the doc to remove myxattr
- assigning a channel based on the non-existence of myxattr
- attempting to write a user xattr named "" (not sure if that's supported by kv).  If so, ensure it's not used if UserXattrKey is undefined

Reply:
@adamcfraser The first two are done as part of the imports PR. I figured it was better to write those tests when I had the ability to properly import them.

Reply:
For the third one:
- Tested this and KV doesn't allow xattrs to have an empty key so we shouldn't run into this.
- GetWithXattr doesn't obtain any xattrs if the UserXattrKey == ""